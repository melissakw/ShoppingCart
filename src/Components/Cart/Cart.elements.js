import styled from "styled-components";

export const Loading = styled.div`
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  min-height: 100%;
  background-color: black;
  color: black;
  opacity: 0.2;
  z-index: 100;
  text-align: center;
  padding-top: 50vh;
`;

export const Table = styled.table`

    border-spacing: 30px;

  @media only screen and (max-width: 600px) {
  table,
  thead,
  tbody,
  th,
  td,
  tr {
    text-align: left;
    display: block;
  }

  thead tr {
    text-align: left;
    position: absolute;
    top: -9999px;
    left: -9999px;
  }

  td {
    /* Makes td act like a "row" */
    border: none;
    position: relative;
    white-space: normal;
    text-align: left;
  }

  td:before {
    width: 45%;
    text-align: left;
    content: attr(data-title);

  }
}
`;

export const CartRow = styled.div`
  display: inline-flex;
  flex-wrap: wrap;
  width: 100%;
  justify-content: space-between;
  flex-direction: row;
`;

export const CartProductImage = styled.img`
  width: 150px;
`;

export const RemoveButton = styled.button`
  background-color: white;
  font-size: 14px;
  border: none;
  padding: 0;
  margin-top: 40px;
  overflow: hidden;
  cursor: pointer;
  opacity: 0.6;
  transition: 0.4s;

  &:hover {
    opacity: 1;
  }
`;

export const CartSummary = styled.div`
  margin-top: 60px;
  margin-right: 60px;
  height: 300px;
  padding: 30px;
  border: 2px lightgray solid;
`;

export const ShipTime = styled.p`
  font-size: 14px;
  font-style: italic;
`;

export const CheckoutButton = styled.button`
  background-color: black;
  color: white;
  width: 100%;
  font-size: 20px;
  border: none;
  padding: 10px;
  margin-top: 10px;
  overflow: hidden;
  cursor: pointer;
  opacity: 0.6;
  transition: 0.4s;

  &:hover {
    opacity: 1;
  }
`;
