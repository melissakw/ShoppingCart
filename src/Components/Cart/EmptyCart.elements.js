import styled from "styled-components";
import { Link } from "react-router-dom";

export const EmptyCartContainer = styled.div`
  padding: 20px;
  text-align: center;
`;

export const ShopLink = styled(Link)`
  color: black;
`;

export const HighlightText = styled.div`
  font-family: "Times New Roman", Times, serif;
  font-size: 34px;
  margin-bottom: 15px;
`;

export const BodyText = styled.div`
  font-size: 16px;
`;
