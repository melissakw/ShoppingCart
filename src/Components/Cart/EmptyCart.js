import React from "react";
import {
  EmptyCartContainer,
  ShopLink,
  HighlightText,
  BodyText
} from "./EmptyCart.elements";

const EmptyCart = () => {
  return (
    <EmptyCartContainer>
      <HighlightText>Your cart is empty</HighlightText>
      <ShopLink to="/">
        <BodyText>Find something you love to fill it up</BodyText>
      </ShopLink>
    </EmptyCartContainer>
  );
};

export default EmptyCart;
