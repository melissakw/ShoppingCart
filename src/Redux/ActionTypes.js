export const ADD_ITEM = "ADD_ITEM";
export const REMOVE_ITEM = "REMOVE_ITEM";
export const EDIT_ITEM_QTY = "EDIT_ITEM_QTY";

export const ADD_TO_CART_LOADING = "ADD_TO_CART_LOADING";
export const ADD_TO_CART_FAILED = "ADD_TO_CART_FAILED";
