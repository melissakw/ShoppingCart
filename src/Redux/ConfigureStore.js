import { createStore, applyMiddleware } from "redux";
import { persistStore, persistCombineReducers } from "redux-persist";
import storage from "redux-persist/lib/storage";
import thunk from "redux-thunk";
import logger from "redux-logger";

import { product } from "./product";

export const ConfigureStore = () => {
  const config = {
    key: "root",
    storage: storage,
    debug: true
  };

  const store = createStore(
    persistCombineReducers(config, {
      product
    }),
    applyMiddleware(thunk, logger)
  );

  const persistor = persistStore(store);
  return { persistor, store };
};
