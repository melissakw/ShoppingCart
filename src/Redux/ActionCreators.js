import * as ActionTypes from "./ActionTypes";

export const postToCart = (product) => (dispatch) => {
  const productWithQty = {
    id: product.id,
    title: product.title,
    price: product.price,
    image: product.image,
    quantity: 1, //defualt quantity to 1 when adding item to cart
    stock: 10
  };
  dispatch(addToCart(productWithQty));
};

export const addToCart = (product) => ({
  type: ActionTypes.ADD_ITEM,
  payload: product
});

export const removeItem = (id) => ({
  type: ActionTypes.REMOVE_ITEM,
  payload: id
});

export const editItem = (product, editedQuantity) => ({
  type: ActionTypes.EDIT_ITEM_QTY,
  product: product,
  quantity: editedQuantity
});

export const addToCartLoading = () => ({
  type: ActionTypes.ADD_TO_CART_LOADING
});

export const addToCartFailed = (errmess) => ({
  type: ActionTypes.ADD_TO_CART_FAILED,
  payload: errmess
});
