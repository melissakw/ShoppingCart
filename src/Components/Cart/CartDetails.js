import React from "react";
import { useDispatch } from "react-redux";
import { removeItem, editItem } from "../../Redux/ActionCreators";
import {
  Table,
  CartProductImage,
  RemoveButton
} from "./Cart.elements";

const CartDetails = ({ productsInCart, setIsLoadingChanges }) => {
  const dispatch = useDispatch();

  return (
    <Table>
      <thead>
        <tr>
          <th></th>
          <th>Item</th>
          <th>Item Price</th>
          <th>Item Quantity</th>
        </tr>
      </thead>
      <tbody>
        {productsInCart.map((productInCart) => {
          const stockQty = Array(productInCart.stock)
            .fill()
            .map((_, i) => i + 1);

          return (
            <tr key={productInCart.id}>
              <td>
                <CartProductImage
                  src={productInCart.image}
                  alt={productInCart.title}
                />
              </td>
              <td>
                <p>{productInCart.title}</p>
              </td>
              <td>
                <p>${productInCart.price}</p>
              </td>
              <td>
                <div>
                  <select
                    name="quantity"
                    id="quantity"
                    defaultValue={productInCart.quantity}
                    onChange={(e) => {
                      setIsLoadingChanges(true);
                      dispatch(editItem(productInCart, e.target.value));
                      setTimeout(() => setIsLoadingChanges(false), 1800);
                    }}
                    style={{ marginTop: "58px" }}
                  >
                    {stockQty.map((num) => {
                      return <option key={num} value={num}>{num}</option>;
                    })}
                  </select>
                </div>
                <div>
                  <RemoveButton
                    onClick={() => {
                      setIsLoadingChanges(true);
                      dispatch(removeItem(productInCart.id));
                      setTimeout(() => setIsLoadingChanges(false), 1800);
                    }}
                  >
                    remove
                  </RemoveButton>
                </div>
              </td>
            </tr>
          );
        })}
      </tbody>
    </Table>
  );
};

export default CartDetails;
