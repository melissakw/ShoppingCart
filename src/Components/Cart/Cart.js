import React, { useState } from "react";
import { useSelector } from "react-redux";
import HeaderBar from "../HeaderBar/HeaderBar";
import EmptyCart from "./EmptyCart";
import CartDetails from "./CartDetails";
import CircularProgress from "@material-ui/core/CircularProgress";
import { AppContainer } from "../App.elements";
import {
  Loading,
  CartRow,
  CartSummary,
  ShipTime,
  CheckoutButton
} from "./Cart.elements";

const Cart = () => {
  const productsInCart = useSelector((state) => state.product.cartItems);
  const totalCartPrice =
    Math.round(
      productsInCart.reduce((a, c) => a + c.price * c.quantity, 0) * 100
    ) / 100;
  const numOfItemsInCart = useSelector((state) =>
    state.product.cartItems.reduce((a, c) => a + +c.quantity, 0)
  );
  const [isLoadingChanges, setIsLoadingChanges] = useState(false);

  return (
    <AppContainer>
      <HeaderBar />
      <h1>Shopping Cart</h1>
      <div>{numOfItemsInCart === 0 && <EmptyCart />}</div>
      <div>
        {isLoadingChanges && (
          <Loading>
            <CircularProgress />
          </Loading>
        )}
      </div>

      <CartRow>
        <CartDetails
          productsInCart={productsInCart}
          setIsLoadingChanges={setIsLoadingChanges}
        />
        <CartSummary>
          <h2>Order Summary</h2>
          <p>Subtotal: ${totalCartPrice}</p>
          <p>Shipping: TBD</p>
          <p>Estimated Tax: $0.00</p>
          <p>Total: ${totalCartPrice}</p>
          <ShipTime>Estimated Delivery: Feb 16-24</ShipTime>
          {/*date for demo only. Actual date to be calculated*/}
          <CheckoutButton>Checkout</CheckoutButton>
        </CartSummary>
      </CartRow>
    </AppContainer>
  );
};

export default Cart;
