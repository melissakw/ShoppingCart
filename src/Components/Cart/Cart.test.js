import React from 'react';
import Cart from './Cart';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from "react-redux";
import { ConfigureStore } from "../../Redux/ConfigureStore";
import { BrowserRouter } from 'react-router-dom';

test("renders Cart correctly", () => {
	const { store } = ConfigureStore();
	const tree = renderer.create(
		<Provider store={store}>
			<BrowserRouter>
				<Cart />
			</BrowserRouter>
		</Provider>
	).toJSON();
	expect(tree).toMatchSnapshot();
});
