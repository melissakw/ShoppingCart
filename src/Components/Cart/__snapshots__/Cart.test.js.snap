// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Cart correctly 1`] = `
<div
  className="sc-bdfBwQ olWrH"
>
  <div
    className="sc-gsTCUz jLYoCy"
  >
    <a
      className="sc-dlfnbm ewOWZD"
      href="/"
      onClick={[Function]}
    >
      <h3>
        THE SHOP
      </h3>
    </a>
    <div
      className="sc-hKgILt dxqMOT"
    >
      <a
        href="/cart"
        onClick={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          data-testid="shopping-icon"
          focusable="false"
          style={
            Object {
              "color": "black",
            }
          }
          viewBox="0 0 24 24"
        >
          <path
            d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zM1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2H1zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
          />
        </svg>
      </a>
      <div
        className="sc-eCssSg zpqXE badge badge-dark"
      >
        0
      </div>
    </div>
  </div>
  <h1>
    Shopping Cart
  </h1>
  <div>
    <div
      className="sc-gKsewC fNySHJ"
    >
      <div
        className="sc-fubCfw hZYuyy"
      >
        Your cart is empty
      </div>
      <a
        className="sc-iBPRYJ gccwaF"
        href="/"
        onClick={[Function]}
      >
        <div
          className="sc-pFZIQ gqSudd"
        >
          Find something you love to fill it up
        </div>
      </a>
    </div>
  </div>
  <div />
  <div
    className="sc-iqHYGH hkfElO"
  >
    <table
      className="sc-kEjbxe jWUUgM"
    >
      <thead>
        <tr>
          <th />
          <th>
            Item
          </th>
          <th>
            Item Price
          </th>
          <th>
            Item Quantity
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
    <div
      className="sc-bqyKva ffJEES"
    >
      <h2>
        Order Summary
      </h2>
      <p>
        Subtotal: $
        0
      </p>
      <p>
        Shipping: TBD
      </p>
      <p>
        Estimated Tax: $0.00
      </p>
      <p>
        Total: $
        0
      </p>
      <p
        className="sc-kstrdz dgWiar"
      >
        Estimated Delivery: Feb 16-24
      </p>
      <button
        className="sc-hBEYos hCWKSF"
      >
        Checkout
      </button>
    </div>
  </div>
</div>
`;
